src/VisualStudio/Core/Def/Implementation/Remote/SolutionChecksumUpdater.cs

```diff
@@ -73,8 +75,15 @@ protected override void PauseOnGlobalOperation()
             CancelAndDispose(previousCancellationSource);
         }
 
-        protected override Task WaitAsync(CancellationToken cancellationToken)
-            => _event.WaitAsync(cancellationToken);
+        protected override async Task WaitAsync(CancellationToken cancellationToken)
+        {
+            var currentToken = await _workQueue.DequeueAsync(cancellationToken).ConfigureAwait(false);
+            lock (_gate)
+            {
+                Contract.ThrowIfFalse(_currentToken is null);
+                _currentToken = currentToken;
+            }
```

Comment:
tbh, i've stared at thsi a bunch, and i have no feeling on what's happening here or why the new code is more correct.  Can you write a non-brief explanation of what this *class* is trying to accomplish, how tehse two fields play into it, and why the new approach is 'correct'?

Reply:
https://github.com/dotnet/roslyn/issues/46255#issuecomment-664043259

Reply:
![image](https://user-images.githubusercontent.com/1408396/88574098-9fe96900-cff6-11ea-832c-17ac07279a1d.png)